anomalib/models/cfa/torch_model.py

```diff
@@ -0,0 +1,387 @@
+
+        gaussian_blur = GaussianBlur2d(sigma=4).to(score.device)
+        anomaly_map = gaussian_blur(anomaly_map)  # pylint: disable=not-callable
+        return anomaly_map
+
+    def forward(self, input_tensor: Tensor) -> Tensor:
+        """Forward pass.
+
+        Args:
+            input_tensor (Tensor): Input tensor.
+
+        Raises:
+            ValueError: When the memory bank is not initialized.
+
+        Returns:
+            Tensor: Loss or anomaly map depending on the train/eval mode.
+        """
+        if self.memory_bank.ndim == 0:
+            raise ValueError("Memory bank is not initialized. Run `initialize_centroid` method first.")
+
+        self.feature_extractor.eval()
+        with torch.no_grad():
+            features = self.feature_extractor(input_tensor)
+            features = list(features.values())
+
+        if hasattr(self, "scale") is False:
+            self.scale = features[0].size()[-2:]
+
+        target_features = self.descriptor(features)
+        distance = self.compute_distance(target_features)
+
+        if self.training:
+            output = self.compute_loss(distance)
+        else:
+            score = self.compute_score(distance)
+            output = self.compute_anomaly_map(score)
+
+        return output
+
+
+class Descriptor(nn.Module):
+    """Descriptor module."""
+
+    def __init__(self, gamma_d: int, backbone: str) -> None:
+        super().__init__()
+
+        self.backbone = backbone
+        if self.backbone not in SUPPORTED_BACKBONES:
+            raise ValueError(f"Supported backbones are {SUPPORTED_BACKBONES}. Got {self.backbone} instead.")
+
+        # TODO: Automatically infer the number of dims
+        backbone_dims = {"vgg19_bn": 1280, "resnet18": 448, "wide_resnet50_2": 1792, "efficientnet_b5": 568}
+        dim = backbone_dims[backbone]
+        out_channels = 2 * dim // gamma_d if backbone == "efficientnet_b5" else dim // gamma_d
+
+        self.layer = CoordConv2d(in_channels=dim, out_channels=out_channels, kernel_size=1)
+
+    def forward(self, features: Union[List[Tensor], Dict[str, Tensor]]) -> Tensor:
+        """Forward pass."""
+        if isinstance(features, dict):
+            features = list(features.values())
+
+        patch_features: Optional[Tensor] = None
+        for i in features:
+            i = F.avg_pool2d(i, 3, 1, 1) / i.size(1) if self.backbone == "efficientnet_b5" else F.avg_pool2d(i, 3, 1, 1)
+            patch_features = (
+                i
+                if patch_features is None
+                else torch.cat((patch_features, F.interpolate(i, patch_features.size(2), mode="bilinear")), dim=1)
+            )
+
+        target_oriented_features = self.layer(patch_features)
+        return target_oriented_features
+
+
+class CoordConv2d(nn.Conv2d):
+    """CoordConv layer as in the paper.
+
+    Link to the paper: https://arxiv.org/abs/1807.03247
+    Link to the PyTorch implementation: https://github.com/walsvid/CoordConv
```

Comment:
I guess we should include the copyright notice of the original author somewhere (for example at the top of the file)

Reply:
Done